kinematics_interface_kdl/src/kinematics_interface_kdl.cpp

```diff
@@ -20,232 +20,229 @@ namespace kinematics_interface_kdl {
-                                               const std::string &link_name,
-                                               std::vector<double> &transform_vec) {
-    // verify inputs
-    if (!verify_initialized() || !verify_joint_vector(joint_pos) || !verify_link_name(link_name) ||
-        !verify_transform_vector(transform_vec)) {
-      return false;
-    }
-
-    // get joint array
-    update_joint_array(joint_pos);
+  // get joint array
+  update_joint_array(joint_pos);
 
-    // reset transform_vec
-    memset(transform_vec.data(), 0, 16 * sizeof(double));
+  // reset transform_vec
+  transform.setIdentity();
 
-    // special case: since the root is not in the robot tree, need to return identity transform
-    if (link_name == root_name_) {
-      transform_vec[0] = 1.0;
-      transform_vec[5] = 1.0;
-      transform_vec[10] = 1.0;
-      transform_vec[15] = 1.0;
-      return true;
-    }
-
-    // create forward kinematics solver
-    fk_pos_solver_->JntToCart(q_, frame_, link_name_map_[link_name]);
-    double tmp[] = {frame_.p.x(), frame_.p.y(), frame_.p.z()};
-    // KDL::Rotation stores data in row-major format. e.g Xx, Yx, Zx, Xy... = data index at 0, 1, 2, 3, 4...
-    for (int r = 0; r < 3; r++) {
-      for (int c = 0; c < 3; c++) {
-        transform_vec[r + 4 * c] = frame_.M.data[3 * r + c];
-      }
-      transform_vec[4 * 3 + r] = tmp[r];
-    }
-    transform_vec[15] = 1.0;
+  // special case: since the root is not in the robot tree, need to return identity transform
+  if (link_name == root_name_)
+  {
     return true;
   }
 
-  bool KDLKinematics::update_joint_array(const std::vector<double> &joint_pos) {
-    memcpy(q_.data.data(), joint_pos.data(), joint_pos.size() * sizeof(double));
-    return true;
+  // create forward kinematics solver
+  fk_pos_solver_->JntToCart(q_, frame_, link_name_map_[link_name]);
+  double tmp[] = {frame_.p.x(), frame_.p.y(), frame_.p.z()};
+  // KDL::Rotation stores data in row-major format. e.g Xx, Yx, Zx, Xy... = data index at 0, 1, 2, 3, 4...
+  for (int r = 0; r < 3; r++)
+  {
+    for (int c = 0; c < 3; c++)
+    {
+      transform(r + 4 * c) = frame_.M.data[3 * r + c];
+    }
+    transform(4 * 3 + r) = tmp[r];
   }
+  return true;
+}
 
-  bool KDLKinematics::verify_link_name(const std::string &link_name) {
-    if (link_name == root_name_) {
-      return true;
-    }
-    if (link_name_map_.find(link_name) == link_name_map_.end()) {
-      std::string links;
-      for (auto i = 0u; i < chain_.getNrOfSegments(); i++) {
-        links += "\n" + chain_.getSegment(i).getName();
-      }
-      RCLCPP_ERROR(LOGGER, "The link %s was not found in the robot chain. Available links are: %s",
-                   link_name.c_str(), links.c_str());
-      return false;
-    }
-    return true;
-  }
+bool KDLKinematics::update_joint_array(const Eigen::Matrix<double, Eigen::Dynamic, 1> &joint_pos)
+{
+  memcpy(q_.data.data(), joint_pos.data(), joint_pos.size() * sizeof(double));
```

Comment:
Hopefully you can simply do something like this instead of memcpy now  (I hardly ever see the usage of memcpy).

```suggestion
  q_ = joint_pos;
```

Reply:
Done